Use `py_literal` for parsing

sudo-carson opened the pull request:
Uses [`py_literal`](https://crates.io/crates/py_literal) for parsing, removing the dependency on nom 3. Fixes #56.

ExpHP commented on src/header.rs:
```diff
@@ -551,7 +473,7 @@ mod tests {
     #[test]
     fn errors_when_record_list_entry_contains_too_few_items() {
-        let faulty_list = parse("[('a')]");
+        let faulty_list = parse("[('a',)]");
```
ah, whoops!  Nice catch.

ExpHP commented:
Thanks, I like this change in spirit.  It has always bothered me how the NPY spec specifies the header as a "Python literal expression of a dictionary," since a truly correct parser for this would be a fair amount of work.

I'll take a more detailed look when I have more time (hopefully tonight).  I'll especially want to look at the number of recursive dependencies; the handrolled parser wasn't great, but if there's one good thing about `nom`, it's the lack of recursive dependencies.

ExpHP commented:
Okay, here's an analysis of the dependency changes:

**Build times for `--no-default-features`**
(i7-12700K, 8x2 threads at 3.61GHz)
| | Before | After |
| --- | --- | --- |
| Debug | 1.727s | 4.929s |
| Release | 2.451s | 6.039s |

```
$ rm -rf target && time cargo build --no-default-features
    Updating crates.io index
   Compiling libc v0.2.131
   Compiling byteorder v1.4.3
   Compiling memchr v1.0.2
   Compiling nom v3.2.1
   Compiling npyz v0.7.3 (/home/exp/asd/clone/npyz)
    Finished dev [unoptimized + debuginfo] target(s) in 1.70s
```

```
$ rm -rf target && time cargo build --no-default-features
    Updating crates.io index
   Compiling proc-macro2 v1.0.43
   Compiling autocfg v1.1.0
   Compiling unicode-ident v1.0.3
   Compiling quote v1.0.21
   Compiling syn v1.0.99
   Compiling thiserror v1.0.38
   Compiling ucd-trie v0.1.5
   Compiling once_cell v1.17.0
   Compiling byteorder v1.4.3
   Compiling num-traits v0.2.15
   Compiling num-integer v0.1.45
   Compiling num-bigint v0.4.3
   Compiling num-complex v0.4.2
   Compiling thiserror-impl v1.0.38
   Compiling pest v2.5.4
   Compiling pest_meta v2.5.4
   Compiling pest_generator v2.5.4
   Compiling pest_derive v2.5.4
   Compiling py_literal v0.4.0
   Compiling npyz v0.7.3 (/home/exp/asd/clone/npyz)
    Finished dev [unoptimized + debuginfo] target(s) in 4.90s
```

The Cargo.lock also grows approximately by a factor of 2.

Looking at the dependencies, I can group them as follows:

- `syn`/`quote`/`proc_macro2`/`unicode-ident`.  These days it's tough to keep proc macros out of most crates and it's frankly a miracle that npyz has been able to keep these deps optional to this day.
- `thiserror`.  Though never truly necessary, I think `thiserror` is a reasonably cheap dependency to have in any dep tree that already uses proc macros, and I generally encourage its use both in library and binary code.
- Various `num-*` crates.  I'm not a fan of the the num traits, but most users of npyz will also likely be using `ndarray` which also depends on these.
- `pest`.  I don't fault py-literal for using a parser generator, and this looks pretty popular and well-maintained.
- `py-literal` itself.  It's maintained by bluss and Jim Turner, two pillars of the rust community.

---

While I do have a bit of pride in npyz's low dependency count at the moment, the current parser is just plain *incorrect.*  All of these dependencies seem acceptable to have in the sort of projects where `npyz` is most likely to be used (high-level or numerical computing code), so I do not consider the additional deps to be a blocker.

I will review the changes in detail shortly.

ExpHP commented on src/header.rs:
```diff
@@ -183,15 +186,18 @@ fn extract_full_array_shape(mut dtype: &DType) -> (Vec<u64>, &DType) {
 pub(crate) fn convert_value_to_shape(field: &Value) -> io::Result<Vec<u64>> {
     if let Value::List(ref lengths) = *field {
         lengths.iter().map(convert_value_to_shape_integer).collect()
+    } else if let Value::Tuple(ref lengths) = *field {
```
Minor nit, I would either change this (and the other places where `else if let Value::Tuple` was added) to a `match` or, even more preferrably, since this logic occurs in many places, I would suggest factoring it out into a private `fn convert_value_to_sequence(&Value) -> &[Value]`

ExpHP commented on src/header.rs:
```diff
@@ -154,15 +154,18 @@ fn convert_tuple_to_record_field(tuple: &[Value]) -> io::Result<Field> {
-            Ok(number as u64)
+    if let Value::Integer(number) = number {
+        let parts = number.to_u64_digits().1;
+        if !parts.is_empty() && parts[0] > 0 {
```
Does this comparison not trivially succeed?  I would expect we should be testing the `num_bigint::Sign` returned by `to_u64_digits`.

I believe we also should check that the number is not larger than a `u64`, by checking the length of the digits.

ExpHP commented on src/header.rs:
```diff
@@ -183,15 +186,18 @@ fn extract_full_array_shape(mut dtype: &DType) -> (Vec<u64>, &DType) {
-            Ok(number as u64)
+    if let Value::Integer(number) = number {
+        let parts = number.to_u64_digits().1;
+        if !parts.is_empty() && parts[0] > 0 {
```
Here is this logic again.  Especially given the other comments on it above, I would suggest factoring it out; perhaps rename this function to `convert_value_to_nonnegative_integer`, and then implement `convert_value_to_positive_integer` in terms of it.

ExpHP commented on src/header.rs:
```diff
@@ -232,18 +229,17 @@ pub(crate) fn read_header(r: &mut dyn io::Read) -> io::Result<Value> {
+    let without_newline = match bytes.split_last() {
+        Some((&b'\n', rest)) => rest,
+        Some(_) | None => return Err(invalid_data("missing newline")),
+    };
```
How come a newline was added on line 90 if it is being stripped here?

ExpHP commented on src/header.rs:
```diff
@@ -232,18 +229,17 @@ pub(crate) fn read_header(r: &mut dyn io::Read) -> io::Result<Value> {
+        Some(_) | None => return Err(invalid_data("missing newline")),
+    };
+    let as_str = std::str::from_utf8(without_newline).map_err(|_| invalid_data("could not parse utf-8"))?;
+    match as_str.parse() {
```
Minor format nit:  Now that IResult is gone, this whole `match` can be replaced with `.map_err`.

ExpHP commented on Cargo.toml:
```diff
@@ -41,10 +41,10 @@ default-features = false
+zip = { version = "0.6" }  # NOTICE: also in dependencies
 
 [features]
 default = []
```
No appropriate line to put this on the diff so I'll just put this here to remind myself:  The documention of the `derive` feature in `lib.rs` should be changed to no longer say that it adds build-time dependencies on `syn`/`quote`/etc.  (you can fix this or I'll fix it after)

ExpHP commented on src/header.rs:
```diff
@@ -585,8 +507,6 @@ mod tests {
     }
 
```
My comments about negative/large integers above have made me realize that we are missing tests for negative integers and integers larger than u64.

ExpHP reviewed (changes requested):
Thank you for your pull request!  I do have some questions and minor nits before merging.

sudo-carson commented:
Addressed your comments in https://github.com/ExpHP/npyz/pull/60/commits/e2dddbce5239eb66dbebc1e95e5b12e776a9c397 and https://github.com/ExpHP/npyz/pull/60/commits/7be763b0504a93e7f0038482a3fb1ecdd5c4716f, please let me know if there's anything else I should change.

ExpHP reviewed (approved):
lgtm

ExpHP commented:
npyz 0.7.4 has been released, including this change and some fixes for structured arrays (zero-size arrays and funny characters in member names)
